sdk/framework/openapi.go

```diff
@@ -575,84 +565,188 @@ func specialPathMatch(path string, specialPaths []string) bool {
-	})
-
-	// Initialize paths with the original pattern or the halves of an
-	// alternation, which is also present in some patterns.
-	matches := altRe.FindAllStringSubmatch(pattern, -1)
-	if len(matches) > 0 {
-		paths = []string{matches[0][1], matches[0][2]}
-	} else {
-		paths = []string{pattern}
+	paths, err := collectPathsFromRegexpAST(rx)
+	if err != nil {
+		// Pattern cannot be transformed into sensible OpenAPI paths, so drop it entirely.
+		// See lengthier comment inside collectPathsFromRegexpASTInternal below for more detailed discussion.
+		return nil
 	}
 
-	// Expand all optional regex elements into two paths. This approach is really only useful up to 2 optional
-	// groups, but we probably don't want to deal with the exponential increase beyond that anyway.
-	for i := 0; i < len(paths); i++ {
-		p := paths[i]
-
-		// match is a 2-element slice that will have a start and end index
-		// for the left-most match of a regex of form: (lease/)?
-		match := optRe.FindStringIndex(p)
+	return paths
+}
 
-		if match != nil {
-			// create a path that includes the optional element but without
-			// parenthesis or the '?' character.
-			paths[i] = p[:match[0]] + p[match[0]+1:match[1]-2] + p[match[1]:]
+type pathCollector struct {
+	strings.Builder
+	conditionalSlashAppendedAtLength int
+}
 
-			// create a path that excludes the optional element.
-			paths = append(paths, p[:match[0]]+p[match[1]:])
-			i--
+// collectPathsFromRegexpAST performs a depth-first recursive walk through a regexp AST, collecting an OpenAPI-style
+// path as it goes.
+//
+// Each time it encounters alternation (a|b) or an optional part (a?), it forks its processing to produce additional
+// results, to account for each possibility. Note: This does mean that an input pattern with lots of these regexp
+// features can produce a lot of different OpenAPI endpoints. At the time of writing, the most complex known example is
+//
+//	"issuer/" + framework.GenericNameRegex(issuerRefParam) + "/crl(/pem|/der|/delta(/pem|/der)?)?"
+//
+// in the PKI secrets engine which expands to 6 separate paths.
+//
+// Each named capture group - i.e. (?P<name>something here) - is replaced with an OpenAPI parameter - i.e. {name} - and
+// the subtree of regexp AST inside the parameter is completely skipped.
+func collectPathsFromRegexpAST(rx *syntax.Regexp) ([]string, error) {
+	pathCollectors, err := collectPathsFromRegexpASTInternal(rx, []*pathCollector{{}})
+	if err != nil {
+		return nil, err
+	}
+	paths := make([]string, 0, len(pathCollectors))
+	for _, collector := range pathCollectors {
+		if collector.conditionalSlashAppendedAtLength != collector.Len() {
+			paths = append(paths, collector.String())
 		}
 	}
+	return paths, nil
+}
 
-	// Replace named parameters (?P<foo>) with {foo}
-	var replacedPaths []string
+func collectPathsFromRegexpASTInternal(rx *syntax.Regexp, appendingTo []*pathCollector) ([]*pathCollector, error) {
+	var err error
+
+	switch rx.Op {
```

Comment:
Nit on code organization: it's a little hard to tell that these cases fall through to `return appendingTo, nil`. Perhaps refactoring some of the larger case blocks into their own functions could help. It may be just me not being able to parse these correctly so I'm OK to keep it as is as well.

Reply:
You are not the first person to comment that my preferred code style leans towards big functions, so I did think hard about this. My rationale for doing it this way, is that when you have code that is only used once, in a specific place in the control flow, then factoring it out into a separate function can remove the easily visible evidence that it is in fact only used once in that specific control flow.

Since many of the cases are small, with only OpCapture, OpQuest and OpAlternate being big enough even consider factoring them out, I'd rather not introduce the inconsistency of most Op types being handled in-situ and just a few separated out.

What I can do, and arguably should have done in the first place, is provide a few lines of comment at the top of the switch, to guide someone reading the code top-to-bottom to an overall understanding:

```suggestion
	// Depending on the type of this regexp AST node (its Op, i.e. operation), figure out whether it contributes any
	// characters to the URL path, and whether we need to recurse through child AST nodes.
	//
	// Each element of the appendingTo slice tracks a separate path, defined by the alternatives chosen when traversing
	// the | and ? conditional regexp features, and new elements are added as each of these features are traversed.
	//
	// To share this slice across multiple recursive calls of this function, it is passed down as a parameter to each
	// recursive call, potentially modified throughout this switch block, and passed back up as a return value at the
	// end of this function - the parent call uses the return value to update its own local variable.
	switch rx.Op {
```